xicor/xicor.py

```diff
@@ -1,89 +1,36 @@
-
-    @property
-    def sample_size(self):
-        return len(self.x)
-
-    @property
-    def x_ordered_rank(self):
-        # PI is the rank vector for x, with ties broken at random
-        # Not mine: source (https://stackoverflow.com/a/47430384/1628971)
-        # random shuffling of the data - reason to use random.choice is that
-        # pd.sample(frac=1) uses the same randomizing algorithm
-        len_x = len(self.x)
-        randomized_indices = np.random.choice(np.arange(len_x), len_x, replace=False)
-        randomized = [self.x[idx] for idx in randomized_indices]
-        # same as pandas rank method 'first'
-        rankdata = ss.rankdata(randomized, method="ordinal")
-        # Reindexing based on pairs of indices before and after
-        unrandomized = [
-            rankdata[j] for i, j in sorted(zip(randomized_indices, range(len_x)))
-        ]
-        return unrandomized
-
-    @property
-    def y_rank_max(self):
-        # f[i] is number of j s.t. y[j] <= y[i], divided by n.
-        return ss.rankdata(self.y, method="max") / self.sample_size
-
-    @property
-    def g(self):
-        # g[i] is number of j s.t. y[j] >= y[i], divided by n.
-        return ss.rankdata([-i for i in self.y], method="max") / self.sample_size
-
-    @property
-    def x_ordered(self):
-        # order of the x's, ties broken at random.
-        return np.argsort(self.x_ordered_rank)
-
-    @property
-    def x_rank_max_ordered(self):
-        # Rearrange f according to ord.
-        return [self.y_rank_max[i] for i in self.x_ordered]
-
-    @property
-    def mean_absolute(self):
-        return (
-            np.mean(
-                np.abs(
-                    [
-                        x - y
-                        for x, y in zip(
-                            self.x_rank_max_ordered[0 : (self.sample_size - 1)],
-                            self.x_rank_max_ordered[1 : self.sample_size],
-                        )
-                    ]
-                )
-            )
-            * (self.sample_size - 1)
-            / (2 * self.sample_size)
-        )
-
-    @property
-    def inverse_g_mean(self):
-        return np.mean(self.g * (1 - self.g))
-
-    @property
-    def correlation(self):
-        """xi correlation"""
-        return 1 - self.mean_absolute / self.inverse_g_mean
+        self.x = np.array(list(x))
+        self.y = np.array(list(y))
+        x_tiebroken = self.x.copy()
+        a,ix,c = np.unique(x_tiebroken,return_counts=True,return_inverse=True)
+        f = c[ix] > 1
+        x_tiebroken[f] = x_tiebroken[f] + np.random.rand(f.sum())*1e-6
+        self.x_ordered = np.argsort(x_tiebroken, kind='quicksort')  
+        self.sample_size = len(self.x)
+
+        _,b,c = np.unique(self.y[self.x_ordered],return_counts=True,return_inverse=True)
+        self.f = np.cumsum(c)[b]        
+        _,b,c = np.unique(-self.y[self.x_ordered],return_counts=True,return_inverse=True)
+        self.g = np.cumsum(c)[b]  
```

Comment:
This is quite nice but I think it is possible to make it even faster with a single call to `unique`. E.g. something like this:
```
def _ranks(x):
    _, inverse, counts = np.unique(
        x, return_index=False, return_inverse=True, return_counts=True
    )
    r = np.cumsum(counts)[inverse]
    # Compute l the same way but flip the counts first.
    counts = np.flip(counts)
    inverse = len(counts) - inverse - 1
    l = np.cumsum(counts)[inverse]
    return r, l
```